### PR TITLE
Tensor product matrix creator

### PR DESCRIPTION
**Pull request by peterrum**
depends on #14288 and #14289

### CODE REVIEW

**Comment by kronbichler** [comment]
Can you rebase this so I can have a new look?

**Comment by peterrum** [comment]
> Can you rebase this so I can have a new look?

Done!

File: include/deal.II/lac/tensor_product_matrix.h [hunk 1]
```diff
@@ -486,6 +486,61 @@ class TensorProductMatrixSymmetricSumCollection
...
+  };
+
+  /**
+   * Create 1D mass matrix and 1D derivative matrix for a scalar
+   * consant-coefficient
+   * Laplacian cell for a @p dim dimensional Cartesian cell. Its boundary types
+   * can be specified with @p bids. The cell extend (including the cell extend
+   * of each neighbor) can be specified via @p cell_extend. With @p n_overlap, an
+   * overlap with neighboring cells can be specified. The default value is one,
+   * which correspond to all matrix enties restricted to the cell-local DoFs.
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:513]:
```suggestion
   * which correspond to all matrix entries restricted to the cell-local DoFs.
```

File: include/deal.II/lac/tensor_product_matrix.h [hunk 2]
```diff
@@ -486,6 +486,61 @@ class TensorProductMatrixSymmetricSumCollection
...
+   */
+  template <int dim, typename Number>
+  std::pair<std::array<FullMatrix<Number>, dim>,
+            std::array<FullMatrix<Number>, dim>>
+  create_laplace_tensor_product_matrix(
+    const FiniteElement<1> &                            fe,
+    const Quadrature<1> &                               quadrature,
+    const dealii::ndarray<LaplaceBoundaryType, dim, 2> &bids,
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:521]:
```suggestion
    const dealii::ndarray<LaplaceBoundaryType, dim, 2> &boundary_ids,
```

File: include/deal.II/lac/tensor_product_matrix.h [hunk 3]
```diff
@@ -486,6 +486,61 @@ class TensorProductMatrixSymmetricSumCollection
...
+  /**
+   * Same as above but the boundary IDs are extracted from the given @p cell
+   * and are mapped to the boundary type via the sets @p dbcs and @p nbcs.
+   */
+  template <int dim, typename Number>
+  std::pair<std::array<FullMatrix<Number>, dim>,
+            std::array<FullMatrix<Number>, dim>>
+  create_laplace_tensor_product_matrix(
+    const typename Triangulation<dim>::cell_iterator &cell,
+    const std::set<types::boundary_id> &              dbcs,
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:534]:
```suggestion
    const std::set<types::boundary_id> &              dirichlet_boundaries,
```
and similar for the Neumann case; what would you do in the Robin boundary case?

File: include/deal.II/lac/tensor_product_matrix.h [hunk 4]
```diff
@@ -1246,6 +1301,273 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::
 
 
 
+namespace TensorProductMatrixCreator
+{
+  namespace internal
+  {
+    template <typename Number>
+    std::tuple<FullMatrix<Number>, FullMatrix<Number>, bool>
+    create_referece_mass_and_striffness_matrices(
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:1310]:
```suggestion
    create_reference_mass_and_stiffness_matrices(
```

File: include/deal.II/lac/tensor_product_matrix.h [hunk 5]
```diff
@@ -1246,6 +1301,273 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::
...
+
+
+  template <int dim, typename Number>
+  std::pair<std::array<FullMatrix<Number>, dim>,
+            std::array<FullMatrix<Number>, dim>>
+  create_laplace_tensor_product_matrix(
+    const FiniteElement<1> &                            fe,
+    const Quadrature<1> &                               quadrature,
+    const dealii::ndarray<LaplaceBoundaryType, dim, 2> &bids,
+    const dealii::ndarray<double, dim, 3> &             cell_extend,
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:1372]:
```suggestion
    const dealii::ndarray<double, dim, 3> &             cell_extent,
```
extend = verb, extent = noun

File: include/deal.II/lac/tensor_product_matrix.h [hunk 6]
```diff
@@ -1246,6 +1301,273 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::
...
+              {
+                // left NBC -> nothing to do
+              }
+            else
+              {
+                AssertThrow(false, ExcNotImplemented());
+              }
+          }
+
+        // reight neighbor or right boundary
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:1450]:
```suggestion
        // right neighbor or right boundary
```

File: include/deal.II/lac/tensor_product_matrix.h [hunk 7]
```diff
@@ -1246,6 +1301,273 @@ TensorProductMatrixSymmetricSumCollection<dim, Number, n_rows_1d>::
...
+                // right NBC -> nothing to do
+              }
+            else
+              {
+                AssertThrow(false, ExcNotImplemented());
+              }
+          }
+
+        Ms[d] = M;
+        Ks[d] = K;
```
Comment (kronbichler) [inline include/deal.II/lac/tensor_product_matrix.h:1484]: Would it make sense to create a reference to `Ms[d]` and `Ks[d]` rather that copying here in the end?

**Review by kronbichler — COMMENTED** [review]
I think this makes sense, but I do not think this should reside in the `lac` subfolder. How about moving it to `FETools` or something like that?

**Comment by peterrum** [comment]
> I think this makes sense, but I do not think this should reside in the lac subfolder. 

I have moved it to the `numerics` foler.

File: include/deal.II/lac/tensor_product_matrix.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/lac/tensor_product_matrix.h]: I'll need to think about this.

File: include/deal.II/numerics/tensor_product_matrix_creator.h [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+#ifndef dealii_tensor_product_matrix_creator_h
+#define dealii_tensor_product_matrix_creator_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/fe/fe.h>
+
+#include <deal.II/grid/tria.h>
```
Comment (kronbichler) [inline include/deal.II/numerics/tensor_product_matrix_creator.h:24]: I guess it gets transitively included, but I would prefer if we could also `#include <deal.II/base/quadrature.h>`

**Review by kronbichler — APPROVED** [review]
This looks good in my opinion. I think one question I would have for the future is whether there is a way to avoid re-computing the same reference mass and stiffness matrices, as we might call this function many times with just the `cell_extent` and `boundary_ids` differing, but the same underlying 1D matrices.

**Review by kronbichler — APPROVED** [review]
Do you plan tests for this? I think we should move forward here, so we can do it elsewhere. But you should rebase to get the CI going, not sure what happened.

**Pull request closed without merging**